File: Common/EncryptUnit.pas
unit EncryptUnit;

interface
uses
  Windows, SysUtils, Classes;
function DecryStrHex(StrHex: string): string;
function EncryStrHex(Str: string): string;

function DecryScript(Str: string): string;
function EncryScript(Str: string): string;

function DecryString(Str: string): string;
function EncryString(Str: string): string;

function EncryBuffer(Buf: PChar; Bufsize: Integer): string;
procedure DecryBuffer(Src: string; Buf: PChar; Bufsize: Integer);



function EncryBufferK(Buf: PChar; Bufsize: Integer; Key: string): string;
procedure DecryBufferK(Src: string; Buf: PChar; Bufsize: Integer; Key: string);
function zEncryBufferK(Buf: PChar; Bufsize: Integer; Key: string): string;
procedure zDecryBufferK(Src: string; Buf: PChar; Bufsize: Integer; Key: string);
function EncryStringK(const Src, Key: string): string;
function DecryStringK(const Src, Key: string): string;

function zEncryStringK(const Src, Key: string): string;
function zDecryStringK(const Src, Key: string): string;

function EncryBufferA(Buf: PChar; Bufsize: Integer): string; overload;
procedure EncryBufferA(Buf: PChar; Bufsize: Integer; OutBuf: PChar); overload;
procedure DecryBufferA(Src: string; Buf: PChar; Bufsize: Integer); overload;
procedure DecryBufferA(Source: PChar; SourceSize: Integer; Outdata: PChar); overload;

function DecryBufferA(Buf: PChar; Bufsize: Integer): string; overload;
function DecryBufferA(Src: string): string; overload;

function GetKeyValue(Value: Integer): string; overload;
function GetKeyValue: Integer; overload;

function GetRandomValue(Value: Integer): string; overload;
function GetRandomValue: Integer; overload;

implementation
uses ZlibEx, EDcode, UnitDes, Base64;
const
  EncryKey = 20120101;
{
Index:0 Chr(Index):48
Index:1 Chr(Index):49
Index:2 Chr(Index):50
Index:3 Chr(Index):51
Index:4 Chr(Index):52
Index:5 Chr(Index):53
Index:6 Chr(Index):54
Index:7 Chr(Index):55
Index:8 Chr(Index):56
Index:9 Chr(Index):57

Index:0 B64[I]:65 Chr(B64[I]):A
Index:1 B64[I]:66 Chr(B64[I]):B
Index:2 B64[I]:67 Chr(B64[I]):C
Index:3 B64[I]:68 Chr(B64[I]):D
Index:4 B64[I]:69 Chr(B64[I]):E
Index:5 B64[I]:70 Chr(B64[I]):F
Index:6 B64[I]:71 Chr(B64[I]):G
Index:7 B64[I]:72 Chr(B64[I]):H
Index:8 B64[I]:73 Chr(B64[I]):I
Index:9 B64[I]:74 Chr(B64[I]):J
Index:10 B64[I]:75 Chr(B64[I]):K
Index:11 B64[I]:76 Chr(B64[I]):L
Index:12 B64[I]:77 Chr(B64[I]):M
Index:13 B64[I]:78 Chr(B64[I]):N
Index:14 B64[I]:79 Chr(B64[I]):O
Index:15 B64[I]:80 Chr(B64[I]):P
Index:16 B64[I]:81 Chr(B64[I]):Q
Index:17 B64[I]:82 Chr(B64[I]):R
Index:18 B64[I]:83 Chr(B64[I]):S
Index:19 B64[I]:84 Chr(B64[I]):T
Index:20 B64[I]:85 Chr(B64[I]):U
Index:21 B64[I]:86 Chr(B64[I]):V
Index:22 B64[I]:87 Chr(B64[I]):W
Index:23 B64[I]:88 Chr(B64[I]):X
Index:24 B64[I]:89 Chr(B64[I]):Y
Index:25 B64[I]:90 Chr(B64[I]):Z

Index:26 B64[I]:97 Chr(B64[I]):a
Index:27 B64[I]:98 Chr(B64[I]):b
Index:28 B64[I]:99 Chr(B64[I]):c
Index:29 B64[I]:100 Chr(B64[I]):d
Index:30 B64[I]:101 Chr(B64[I]):e
Index:31 B64[I]:102 Chr(B64[I]):f
Index:32 B64[I]:103 Chr(B64[I]):g
Index:33 B64[I]:104 Chr(B64[I]):h
Index:34 B64[I]:105 Chr(B64[I]):i
Index:35 B64[I]:106 Chr(B64[I]):j
Index:36 B64[I]:107 Chr(B64[I]):k
Index:37 B64[I]:108 Chr(B64[I]):l
Index:38 B64[I]:109 Chr(B64[I]):m
Index:39 B64[I]:110 Chr(B64[I]):n
Index:40 B64[I]:111 Chr(B64[I]):o
Index:41 B64[I]:112 Chr(B64[I]):p
Index:42 B64[I]:113 Chr(B64[I]):q
Index:43 B64[I]:114 Chr(B64[I]):r
Index:44 B64[I]:115 Chr(B64[I]):s
Index:45 B64[I]:116 Chr(B64[I]):t
Index:46 B64[I]:117 Chr(B64[I]):u
Index:47 B64[I]:118 Chr(B64[I]):v
Index:48 B64[I]:119 Chr(B64[I]):w
Index:49 B64[I]:120 Chr(B64[I]):x
Index:50 B64[I]:121 Chr(B64[I]):y
Index:51 B64[I]:122 Chr(B64[I]):z
Index:52 B64[I]:60 Chr(B64[I]):<
Index:53 B64[I]:61 Chr(B64[I]):=
Index:54 B64[I]:62 Chr(B64[I]):>
Index:55 B64[I]:64 Chr(B64[I]):@
Index:56 B64[I]:91 Chr(B64[I]):[
Index:57 B64[I]:92 Chr(B64[I]):\
Index:58 B64[I]:93 Chr(B64[I]):]
}


function GetRandomValue(Value: Integer): string;
const
  Random55: array[0..54] of Byte =
    // 0~9 没有0和1
  (50, 51, 52, 53, 54, 55, 56, 57,
   // A~Z 没有73:I 79:O
    65, 66, 67, 68, 69, 70, 71, 72, 74, 75, 76, 77, 78, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90,
    // a~z 没有105:I 108:l 111:o
    97, 98, 99, 100, 101, 102, 103, 104, 106, 107, 109, 110, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122);
var
  btIdx: Byte;
  P: PChar;
begin
  SetLength(Result, 4);
  P := PChar(Result);

  btIdx := LoByte(LoWord(Value));
  P^ := Chr(Random55[btIdx]);
  Inc(P);

  btIdx := HiByte(LoWord(Value));
  P^ := Chr(Random55[btIdx]);
  Inc(P);

  btIdx := LoByte(HiWord(Value));
  P^ := Chr(Random55[btIdx]);
  Inc(P);

  btIdx := HiByte(HiWord(Value));
  P^ := Chr(Random55[btIdx]);
end;

function GetRandomValue: Integer;
begin
  Result := MakeLong(MakeWord(Random(55), Random(55)), MakeWord(Random(55), Random(55)));
end;

function GetKeyValue: Integer;
begin
  Result := MakeLong(MakeWord(Random(58), Random(58)), MakeWord(Random(58), Random(58)));
end;

function GetKeyValue(Value: Integer): string;
const
  B64: array[0..58] of Byte = (65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
    81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108,
    109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 60, 61, 62, 64, 91, 92, 93);
var
  btIdx: Byte;
  P: PChar;
begin
  SetLength(Result, 4);
  P := PChar(Result);

  btIdx := LoByte(LoWord(Value));
  P^ := Chr(B64[btIdx]);
  Inc(P);

  btIdx := HiByte(LoWord(Value));
  P^ := Chr(B64[btIdx]);
  Inc(P);

  btIdx := LoByte(HiWord(Value));
  P^ := Chr(B64[btIdx]);
  Inc(P);

  btIdx := HiByte(HiWord(Value));
  P^ := Chr(B64[btIdx]);
end;

function DecryScript(Str: string): string;
begin
  Result := DecryptStrDes(Str, IntToStr(EncryKey));
end;

function EncryScript(Str: string): string;
begin
  Result := EncryptStrDes(Str, IntToStr(EncryKey));
end;

// /////////////////////////////////////////////////////////  442517066

function DecryString(Str: string): string;
var
  Key: string;
begin
  if Length(Str) > 4 then
  begin
    Key := Str[1] + Str[2] + Str[Length(Str) - 1] + Str[Length(Str)];
    Str := Copy(Str, 3, Length(Str) - 4);
    Result := DecryptStrDes(Base64DecodeStr(DecodeString(Str)), Key);
  end
  else
    Result := '';
end;

function EncryString(Str: string): string;
begin
  if Str <> '' then
    Result := EncryBuffer(@Str[1], Length(Str))
  else
    Result := '';
end;


function EncryBufferA(Buf: PChar; Bufsize: Integer): string;
begin
  SetLength(Result, Bufsize);
  EncryptDes(Buf^, Result[1], Bufsize, IntToStr(EncryKey));
end;

procedure EncryBufferA(Buf: PChar; Bufsize: Integer; OutBuf: PChar);
begin
  EncryptDes(Buf^, OutBuf^, Bufsize, IntToStr(EncryKey));
end;

procedure DecryBufferA(Src: string; Buf: PChar; Bufsize: Integer);
begin
  DecryptDes(Src[1], Buf^, Bufsize, IntToStr(EncryKey));
end;

function DecryBufferA(Buf: PChar; Bufsize: Integer): string;
begin
  SetLength(Result, Bufsize);
  DecryptDes(Buf^, Result[1], Bufsize, IntToStr(EncryKey));
end;

function DecryBufferA(Src: string): string;
begin
  Result := DecryptStrDes(Src, IntToStr(EncryKey));
end;

procedure DecryBufferA(Source: PChar; SourceSize: Integer; Outdata: PChar);
begin
  DecryptDes(Source^, Outdata^, SourceSize, IntToStr(EncryKey));
end;

function EncryBufferK(Buf: PChar; Bufsize: Integer; Key: string): string;
var
  sText: string;
begin
  if Bufsize > 0 then
  begin
    SetLength(sText, Bufsize);
    EncryptDes(Buf^, sText[1], Bufsize, Key);
    Result := EncodeString(Base64EncodeStr(sText));
  end
  else
    Result := '';
end;

function zEncryBufferK(Buf: PChar; Bufsize: Integer; Key: string): string;
var
  sText: string;
  OutBuf: PChar; OutBytes: Integer;
begin
  sText := '';
  if Bufsize > 0 then
  begin
    OutBuf := nil;
    OutBytes := 0;
    try
      CompressBuf(Buf, Bufsize, Pointer(OutBuf), OutBytes);
      SetLength(sText, OutBytes);
      Move(OutBuf^, sText[1], OutBytes);
    finally
      FreeMem(OutBuf);
    end;
    if sText <> '' then
    begin
      SetLength(Result, Length(sText));
      EncryptDes(sText[1], Result[1], Length(sText), Key);
      Result := EncodeStringK(Base64EncodeStr(Result));
    end
    else
      Result := '';
  end
  else
    Result := '';
end;

procedure zDecryBufferK(Src: string; Buf: PChar; Bufsize: Integer; Key: string);
var
  sText, sTemp: string;
  OutBuf: Pointer; OutBytes: Integer;
begin
  if Length(Src) > 0 then
  begin

    sText := Base64DecodeStr(DecodeString(Src));                                                    // DecodeString    DecryStrHex
    SetLength(sTemp, Length(sText));
    DecryptDes(sText[1], sTemp[1], Length(sText), Key);

    if sTemp <> '' then
    begin
      OutBuf := nil;
      OutBytes := 0;
      try
        DecompressBuf(@sTemp[1], Length(sTemp), 0, OutBuf, OutBytes);
        Move(OutBuf^, Buf^, OutBytes);
      finally
        FreeMem(OutBuf);
      end;
    end;
  end;
end;


function EncryStringK(const Src, Key: string): string;
begin
  if Src <> '' then
    Result := EncodeString(Base64EncodeStr(EncryptStrDes(Src, Key)))
  else
    Result := '';
end;

function zEncryStringK(const Src, Key: string): string;
begin
  if Src <> '' then
    Result := zEncodeString(Base64EncodeStr(EncryptStrDes(Src, Key)))
  else
    Result := '';
end;

procedure DecryBufferK(Src: string; Buf: PChar; Bufsize: Integer; Key: string);
var
  sText: string;
begin
  if Length(Src) > 0 then
  begin
    sText := Base64DecodeStr(DecodeStringK(Src));                                                   // DecodeString    DecryStrHex
    DecryptDes(sText[1], Buf^, Bufsize, Key);
  end;
end;



function DecryStringK(const Src, Key: string): string;
begin
  if Src <> '' then
    Result := DecryptStrDes(Base64DecodeStr(DecodeStringK(Src)), Key)
  else
    Result := '';
end;

function zDecryStringK(const Src, Key: string): string;
begin
  if Src <> '' then
    Result := DecryptStrDes(Base64DecodeStr(zDecodeString(Src)), Key)
  else
    Result := '';
end;

function EncryBuffer(Buf: PChar; Bufsize: Integer): string;
const
  B64: array[0..58] of byte = (65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
    81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108,
    109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 60, 61, 62, 64, 91, 92, 93);
var
  I: Integer;
  Key: string;
  P: PChar;
  sText: string;
begin
  if Bufsize > 0 then
  begin
    SetLength(Key, 4);
    P := PChar(Key);
    for I := 1 to 4 do
    begin
      P^ := Chr(B64[Random(58)]);
      Inc(P);
    end;
    SetLength(sText, Bufsize);
    EncryptDes(Buf^, sText[1], Bufsize, Key);
    Result := Key[1] + Key[2] + EncodeString(Base64EncodeStr(sText)) + Key[3] + Key[4];
  end
  else
    Result := '';
end;

procedure DecryBuffer(Src: string; Buf: PChar; Bufsize: Integer);
var
  sText: string;
  Key: string;
begin
  if Length(Src) > 4 then
  begin
    Key := Src[1] + Src[2] + Src[Length(Src) - 1] + Src[Length(Src)];
    Src := Copy(Src, 3, Length(Src) - 4);
    sText := Base64DecodeStr(DecodeString(Src));                                                    // DecodeString    DecryStrHex
    DecryptDes(sText[1], Buf^, Bufsize, Key);
  end;
end;

function EncryStrHex(Str: string): string;
var
  StrResult, TempResult, temp: string;
  I: Integer;
begin
  TempResult := Str;
  StrResult := '';
  for I := 0 to Length(TempResult) - 1 do
  begin
    temp := Format('%x', [Ord(TempResult[I + 1])]);
    if Length(temp) = 1 then temp := '0' + temp;
    StrResult := StrResult + temp;
  end;
  Result := StrResult;
end;

function DecryStrHex(StrHex: string): string;

  function HexToInt(Hex: string): Integer;
  var
    I, Res: Integer;
    Ch: Char;
  begin
    Res := 0;
    for I := 0 to Length(Hex) - 1 do
    begin
      Ch := Hex[I + 1];
      if (Ch >= '0') and (Ch <= '9') then
        Res := Res * 16 + Ord(Ch) - Ord('0')
      else if (Ch >= 'A') and (Ch <= 'F') then
        Res := Res * 16 + Ord(Ch) - Ord('A') + 10
      else if (Ch >= 'a') and (Ch <= 'f') then
        Res := Res * 16 + Ord(Ch) - Ord('a') + 10
      else
        raise Exception.Create('Error: not a Hex String');
    end;
    Result := Res;
  end;

var
  Str, temp: string;
  I: Integer;
begin
  Str := '';
  for I := 0 to Length(StrHex) div 2 - 1 do
  begin
    temp := Copy(StrHex, I * 2 + 1, 2);
    Str := Str + Chr(HexToInt(temp));
  end;
  Result := Str;
end;

end.

File: GXX.CSharp/src/GXX.Core/Crypto/EncryptUnit.cs
using System;
using GXX.Core.Protocol;
using GXX.Core.Rtl;

namespace GXX.Core.Crypto;

/// <summary>
/// EncryptUnit.pas 1:1 转换：DES + Base64 + 6Bit 组合加解密入口。
/// </summary>
public static class EncryptUnit
{
    public const int EncryKey = 20120101;

    private static readonly byte[] B64Table =
    {
        65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80,
        81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108,
        109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 60, 61, 62, 64, 91, 92, 93
    };

    private static readonly byte[] Random55 =
    {
        50, 51, 52, 53, 54, 55, 56, 57,
        65, 66, 67, 68, 69, 70, 71, 72, 74, 75, 76, 77, 78, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90,
        97, 98, 99, 100, 101, 102, 103, 104, 106, 107, 109, 110, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122
    };

    private static readonly Random Rnd = new();

    public static string GetRandomValue(int value)
    {
        var chars = new char[4];
        chars[0] = (char)Random55[DelphiRTL.LoByte(DelphiRTL.LoWord(value))];
        chars[1] = (char)Random55[DelphiRTL.HiByte(DelphiRTL.LoWord(value))];
        chars[2] = (char)Random55[DelphiRTL.LoByte(DelphiRTL.HiWord(value))];
        chars[3] = (char)Random55[DelphiRTL.HiByte(DelphiRTL.HiWord(value))];
        return new string(chars);
    }

    public static int GetRandomValue()
        => DelphiRTL.MakeLong(DelphiRTL.MakeWord(Rnd.Next(55), Rnd.Next(55)), DelphiRTL.MakeWord(Rnd.Next(55), Rnd.Next(55)));

    public static int GetKeyValue()
        => DelphiRTL.MakeLong(DelphiRTL.MakeWord(Rnd.Next(58), Rnd.Next(58)), DelphiRTL.MakeWord(Rnd.Next(58), Rnd.Next(58)));

    public static string GetKeyValue(int value)
    {
        var chars = new char[4];
        chars[0] = (char)B64Table[DelphiRTL.LoByte(DelphiRTL.LoWord(value))];
        chars[1] = (char)B64Table[DelphiRTL.HiByte(DelphiRTL.LoWord(value))];
        chars[2] = (char)B64Table[DelphiRTL.LoByte(DelphiRTL.HiWord(value))];
        chars[3] = (char)B64Table[DelphiRTL.HiByte(DelphiRTL.HiWord(value))];
        return new string(chars);
    }

    public static string DecryScript(string str)
        => UnitDes.DecryptStrDesText(str, DelphiRTL.IntToStr(EncryKey));

    public static string EncryScript(string str)
        => UnitDes.EncryptStrDesText(str, DelphiRTL.IntToStr(EncryKey));

    /// <summary>EncryStrHex：字符串→HEX（GBK 字节）。</summary>
    public static string EncryStrHex(string str)
    {
        byte[] data = EncodingInit.GBK.GetBytes(str ?? "");
        var sb = new System.Text.StringBuilder(data.Length * 2);
        foreach (byte b in data)
            sb.Append(b.ToString("x2"));
        return sb.ToString();
    }

    public static string DecryStrHex(string strHex)
    {
        byte[] data = new byte[strHex.Length / 2];
        for (int i = 0; i < data.Length; i++)
        {
            string temp = DelphiRTL.Copy(strHex, i * 2 + 1, 2);
            data[i] = (byte)HUtil32Of(temp);
        }
        return EncodingInit.GBK.GetString(data);
    }

    private static int HUtil32Of(string hex)
    {
        int res = 0;
        foreach (char ch in hex)
        {
            if (ch >= '0' && ch <= '9') res = res * 16 + ch - '0';
            else if (ch >= 'A' && ch <= 'F') res = res * 16 + ch - 'A' + 10;
            else if (ch >= 'a' && ch <= 'f') res = res * 16 + ch - 'a' + 10;
            else throw new Exception("Error: not a Hex String");
        }
        return res;
    }

    /// <summary>EncryString：随机 4 字符 Key 前缀 + DES 加密 + Base64 + 6Bit。</summary>
    public static string EncryString(string str)
    {
        if (str == "") return "";
        byte[] data = EncodingInit.GBK.GetBytes(str);
        return EncryBuffer(data, data.Length);
    }

    public static string DecryString(string str)
    {
        if (str.Length <= 4) return "";
        string key = str.Substring(0, 2) + str.Substring(str.Length - 2, 2);
        string body = DelphiRTL.Copy(str, 3, str.Length - 4);
        byte[] decoded = EDcode.DecodeString(EncodingInit.GBK.GetBytes(body));
        byte[] b64 = Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded));
        byte[] plain = UnitDes.DecryptStrDes(b64, key);
        return EncodingInit.GBK.GetString(plain);
    }

    public static string EncryBuffer(byte[] buf, int bufsize)
    {
        if (bufsize <= 0) return "";
        var keyChars = new char[4];
        for (int i = 0; i < 4; i++)
            keyChars[i] = (char)B64Table[Rnd.Next(58)];
        string key = new string(keyChars);
        byte[] sText = new byte[bufsize];
        UnitDes.EncryptDes(buf, sText, bufsize, key);
        string b64 = Base64Util.Base64EncodeStr(sText);
        string enc = EncodingInit.GBK.GetString(EDcode.EncodeString(b64));
        return keyChars[0].ToString() + keyChars[1] + enc + keyChars[2] + keyChars[3];
    }

    public static void DecryBuffer(string src, byte[] buf, int bufsize)
    {
        if (src.Length <= 4) return;
        string key = src.Substring(0, 2) + src.Substring(src.Length - 2, 2);
        string body = DelphiRTL.Copy(src, 3, src.Length - 4);
        byte[] decoded = EDcode.DecodeString(EncodingInit.GBK.GetBytes(body));
        byte[] sText = Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded));
        UnitDes.DecryptDes(sText, buf, Math.Min(bufsize, sText.Length), key);
    }

    // ---- K 系列（固定 Key）----

    public static string EncryBufferK(byte[] buf, int bufsize, string key)
    {
        if (bufsize <= 0) return "";
        byte[] sText = new byte[bufsize];
        UnitDes.EncryptDes(buf, sText, bufsize, key);
        string b64 = Base64Util.Base64EncodeStr(sText);
        return EncodingInit.GBK.GetString(EDcode.EncodeString(b64));
    }

    public static void DecryBufferK(string src, byte[] buf, int bufsize, string key)
    {
        if (src.Length == 0) return;
        byte[] decoded = EDcode.DecodeStringK(EncodingInit.GBK.GetBytes(src));
        string sText = EncodingInit.GBK.GetString(Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded)));
        byte[] cipher = EncodingInit.GBK.GetBytes(sText);
        UnitDes.DecryptDes(cipher, buf, Math.Min(bufsize, cipher.Length), key);
    }

    public static string EncryStringK(string src, string key)
    {
        if (src == "") return "";
        byte[] enc = UnitDes.EncryptStrDesBytes(EncodingInit.GBK.GetBytes(src), key);
        string b64 = Base64Util.Base64EncodeStr(enc);
        return EncodingInit.GBK.GetString(EDcode.EncodeString(b64));
    }

    public static string DecryStringK(string src, string key)
    {
        if (src == "") return "";
        byte[] decoded = EDcode.DecodeStringK(EncodingInit.GBK.GetBytes(src));
        byte[] b64 = Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded));
        byte[] plain = UnitDes.DecryptStrDesBytes(b64, key);
        return EncodingInit.GBK.GetString(plain);
    }

    public static string zEncryStringK(string src, string key)
    {
        if (src == "") return "";
        byte[] enc = UnitDes.EncryptStrDesBytes(EncodingInit.GBK.GetBytes(src), key);
        string b64 = Base64Util.Base64EncodeStr(enc);
        return EncodingInit.GBK.GetString(EDcode.zEncodeString(b64));
    }

    public static string zDecryStringK(string src, string key)
    {
        if (src == "") return "";
        byte[] decoded = EDcode.zDecodeString(EncodingInit.GBK.GetBytes(src));
        byte[] b64 = Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded));
        byte[] plain = UnitDes.DecryptStrDes(b64, key);
        return EncodingInit.GBK.GetString(plain);
    }

    public static string zEncryBufferK(byte[] buf, int bufsize, string key)
    {
        if (bufsize <= 0) return "";
        byte[] compressed = Compress.ZlibEx.CompressBuf(buf, bufsize);
        if (compressed == null || compressed.Length == 0) return "";
        byte[] sText = new byte[compressed.Length];
        UnitDes.EncryptDes(compressed, sText, compressed.Length, key);
        string b64 = Base64Util.Base64EncodeStr(sText);
        return EncodingInit.GBK.GetString(EDcode.EncodeStringK(b64));
    }

    public static void zDecryBufferK(string src, byte[] buf, int bufsize, string key)
    {
        if (src.Length == 0) return;
        byte[] decoded = EDcode.DecodeString(EncodingInit.GBK.GetBytes(src));
        string sText = EncodingInit.GBK.GetString(Base64Util.Base64DecodeStr(EncodingInit.GBK.GetString(decoded)));
        byte[] cipher = EncodingInit.GBK.GetBytes(sText);
        byte[] sTemp = new byte[cipher.Length];
        UnitDes.DecryptDes(cipher, sTemp, cipher.Length, key);
        byte[] plain = Compress.ZlibEx.DecompressBuf(sTemp, sTemp.Length);
        if (plain != null && plain.Length > 0)
        {
            int n = Math.Min(plain.Length, bufsize);
            Array.Copy(plain, 0, buf, 0, n);
        }
    }

    // ---- A 系列（默认 EncryKey）----

    public static byte[] EncryBufferA(byte[] buf, int bufsize)
    {
        byte[] result = new byte[bufsize];
        UnitDes.EncryptDes(buf, result, bufsize, DelphiRTL.IntToStr(EncryKey));
        return result;
    }

    public static void EncryBufferA(byte[] buf, int bufsize, byte[] outBuf)
        => UnitDes.EncryptDes(buf, outBuf, bufsize, DelphiRTL.IntToStr(EncryKey));

    public static void DecryBufferA(string src, byte[] buf, int bufsize)
    {
        byte[] cipher = EncodingInit.GBK.GetBytes(src);
        UnitDes.DecryptDes(cipher, buf, Math.Min(bufsize, cipher.Length), DelphiRTL.IntToStr(EncryKey));
    }

    public static byte[] DecryBufferA(byte[] buf, int bufsize)
    {
        byte[] result = new byte[bufsize];
        UnitDes.DecryptDes(buf, result, bufsize, DelphiRTL.IntToStr(EncryKey));
        return result;
    }

    public static string DecryBufferA(string src)
        => UnitDes.DecryptStrDesText(src, DelphiRTL.IntToStr(EncryKey));

    public static void DecryBufferA(byte[] source, int sourceSize, byte[] outdata)
        => UnitDes.DecryptDes(source, outdata, sourceSize, DelphiRTL.IntToStr(EncryKey));

    /// <summary>客户端参数解密（Client.dpr 用）：zDecryBufferK 的结构体便捷版。</summary>
    public static byte[] zDecryStruct(string src, int structSize, string key)
    {
        byte[] buf = new byte[structSize];
        zDecryBufferK(src, buf, structSize, key);
        return buf;
    }
}
